### ISSUE TITLE
Come up with something better than the "try" keyword

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#635]
### Description

Inko supports the syntax `try expression`, which is syntax sugar for a `match` on an `Option` or `Result`. The purpose is to reduce the boilerplate of "unwrap if OK, throw if not" that comes with explicit `match` expressions.

Unfortunately, the syntax doesn't compose well. First, the prefix nature of the syntax means that in calls chains you end up having to nest `try` expressions. Second, while we used to also support `try { expr }`, we currently don't. The result is that for the expression `x.y.z`, where `x.y` and `(x.y).z` may throw, you end up with this:

```inko
try (try x.y).z
```

The second problem is that `try` only supports `Result` and `Option`, and this is implemented in the compiler. Adding support for other types would require something like Rust's [Try](https://doc.rust-lang.org/std/ops/trait.Try.html) trait, which is something I want to avoid given the type soup/complexity this introduces.

Finally, `try` introduces a second way of doing pattern matching, but with a more specific purpose. While we might not be able to avoid this, it would be nice if we'd _only_ have the `match` keyword for pattern matching, with the added benefit of it supporting all types.

Long story short: I'd like to explore alternatives that:

1. Are usable in method call chains
2. Are easy to spot (i.e. no `foo.bar?.baz` nonsense)
3. Support any enum type, without special-casing such types in the compiler (as is done with `Option` and `Result` at the moment)
4. Don't require you to implement a whole bunch of traits and what not

### Related work

- https://doc.rust-lang.org/std/ops/trait.Try.html

**Comment by yorickpeterse** [comment]
Idea 1:

When defining an enum, at most one variant can be marked with `else`, but only if there are exactly two variants, and if the OK/not-else variant wraps exactly 1 argument, like so:

```inko
class enum Result[T, E] {
  case Ok(T)
  case else Error(E)
}
```

The `try` expression then translates to a `match` where the OK case matches against the `case` _without_ `else`, and the error case against the `else` case.

The syntax and rules are a bit wonky, but it would allow for custom types to be used with `try`, without needing to implement a complicated trait.

**Comment by yorickpeterse** [comment]
Idea 2:

Change the semantics of `try expr` such that we allow `try { expr }` again, and such that `try a.b.c` results in `try` being applied to each "step" that returns a `Result` or `Option`, and thus is equivalent to `try (try (try a).b).c`. 

This makes chaining easier, in exchange for error handling being less localized, and potentially confusing diagnostics if different types are returned.

**Comment by thadguidry** [comment]
You can change the "try" keyword with a "step" keyword, or even "do" or "exec".
No matter what, there needs to be some syntax that allows a programmer to define the BLOCK OF STEPS (as well as sub-steps) to try, attempt, do, etc.  Common blocking styles often used to frame **sub-steps** are surrounding enclosure glyphs such as { }, ( ), [ ], < > or indentation levels.  Since **expressions** are often also encapsulated using the { } style, it would make sense to also constrain Inko's universal block style as { } as well.

The idea that FAILURE can happen is most evident in the term "try" and "do".  Which is why "try" is an effective and concise term for "attempt" or "aspire".  "aim" is concise, but misses the mark a little bit  :-) to convey FAILURE might or can happen.

Inko might think of breaking the mold, and allowing defining where "each calls chain is linked" between steps using a single character to delineate each chain link through the use of `#` or `-` etc. as a prefix (like markdown steps), but that may or may not be attractive, and simply unneeded to define chain links, if parsing can be robust enough to tell the compiler where and what delineate's each step in a block (CR, LF, expression block ending, etc.) or "set of called chains"

```
do { }
steps { }
exec { }
aim { }
try { }
 ```

```
do {
- let x = 10
- let y = x + 1
- do {  # the following steps are not "chain linked" with - and thus can run in any order
  Stdout.new.print(z + y)  # <-- failure, z not defined
  let errormsg = "fail"  # <-- set because this expresion was reached first
  Stdout.new.print(errormsg) }  # <-- printed because let errormsg was reached first
}
```

```
# implicit top-down ordering through the block of steps
steps {
  let x = 10
  let y = x + 1
  steps {
    Stdout.new.print(z + y)
    Stdout.new.print(y) }
}
```
